app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -1,13 +1,27 @@
-class SessionsController < ApplicationController
-  def create
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def twitter
+    callback(:twitter)
+  end
+
+  def facebook
+    callback(:facebook)
+  end
+
+  def callback(_provider)
     user = new_or_existing(request.env["omniauth.auth"])
 
-    session[:user_id] = user.id
+    # session[:user_id] = user.id
+    sign_in(user)
     user_params = session.delete(:user_params)
     user.update(user_params) if user_params
     origin = request.env["omniauth.origin"]
     logger.debug "Login request origin #{origin}"
     redirect_to origin || user_path
+  rescue ActiveRecord::RecordInvalid
+    # Could make this check it is actually the email...
+    flash[:alert] = "An account already exists with the email #{request.env["omniauth.auth"].info.email}." \
+                    "Please log in to that account."
+    redirect_to new_user_session_path
```

Comment:
I was wondering about just signing them in if the email matches, but this feels like better security, just in case someone doesn't own the email account that matches their twitter / fb account. But please could you add specs for it